test/cpp/api/functional.cpp

```diff
@@ -1960,3 +1960,63 @@ TEST_F(FunctionalTest, Dropout3d) {
   ASSERT_TRUE(torch::allclose(input_mean, output.mean(), 0.01, 0.05));
   ASSERT_TRUE((input_std <= output.std()).all().item<bool>());
 }
+
+template<c10::ScalarType S, typename T>
+void test_isfinite(const at::Device& device) {
+  const std::vector<T> values = {
+    std::numeric_limits<T>::lowest(),
+    0, 1, 42,
+    std::numeric_limits<T>::min(),
+    std::numeric_limits<T>::max()
+  };
+  for (const auto value : values) {
+    const auto x = torch::full({3, 3}, value, torch::TensorOptions().dtype(S).device(device));
+    ASSERT_TRUE(torch::isfinite(x).all().template item<bool>());
+  }
+  if (std::numeric_limits<T>::has_infinity) {
+    const auto inf = std::numeric_limits<T>::infinity();
+    const auto x = torch::tensor({
+      -inf,
+      std::numeric_limits<T>::lowest(),
+      static_cast<T>(0),
+      static_cast<T>(1),
+      static_cast<T>(42),
+      std::numeric_limits<T>::min(),
+      std::numeric_limits<T>::max(),
+      inf
+    }, torch::TensorOptions().dtype(S).device(device));
+    ASSERT_TRUE(torch::allclose(
+      // torch::allclose does not support comparing torch::kBool
+      torch::isfinite(x).toType(torch::kInt),
+      torch::tensor(
+        {false, true, true, true, true, true, true, false},
+        torch::TensorOptions().device(device)
+      ).toType(torch::kInt)
+    ));
+  }
+  if (std::numeric_limits<T>::has_quiet_NaN) {
+    const auto x = torch::tensor({
+      std::numeric_limits<T>::quiet_NaN()
+    }, torch::TensorOptions().dtype(S).device(device));
+    ASSERT_FALSE(torch::isfinite(x).all().template item<bool>());
+  }
+  if (std::numeric_limits<T>::has_signaling_NaN) {
+    const auto x = torch::tensor({
+      std::numeric_limits<T>::signaling_NaN()
+    }, torch::TensorOptions().dtype(S).device(device));
+    ASSERT_FALSE(torch::isfinite(x).all().template item<bool>());
+  }
+}
+
+TEST_F(FunctionalTest, isfinite) {
+  for (const auto device : {at::Device("cpu"), at::Device("cuda")}) {
```

Comment:
CPU-only tests seem to fail with
```
Nov 19 00:33:35 [ RUN      ] FunctionalTest.isfinite
Nov 19 00:33:35 unknown file: Failure
Nov 19 00:33:35 C++ exception with description "Cannot initialize CUDA without ATen_cuda library. PyTorch splits its backend into two shared libraries: a CPU library and a CUDA library; this error has occurred because you are trying to use some CUDA functionality, but the CUDA library has not been loaded by the dynamic linker for some reason.  The CUDA library MUST be loaded, EVEN IF you don't directly use any symbols from the CUDA library! One common culprit is a lack of -Wl,--no-as-needed in your link arguments; many dynamic linkers will delete dynamic library dependencies if you don't depend on any of their symbols.  You can check if this has occurred by using ldd on your binary to see if there is a dependency on *_cuda.so library. (initCUDA at /var/lib/jenkins/workspace/aten/src/ATen/detail/CUDAHooksInterface.h:63)
Nov 19 00:33:35 frame #0: <unknown function> + 0xcfa83 (0x7fabe4f6ca83 in /var/lib/jenkins/workspace/build/lib/libc10.so)
Nov 19 00:33:35 frame #1: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x1c8 (0x7fabe4f6b118 in /var/lib/jenkins/workspace/build/lib/libc10.so)
Nov 19 00:33:35 frame #2: <unknown function> + 0xb1c616f (0x7fabf03e916f in /var/lib/jenkins/workspace/build/lib/libtorch.so)
Nov 19 00:33:35 frame #3: <unknown function> + 0xb12b6a2 (0x7fabf034e6a2 in /var/lib/jenkins/workspace/build/lib/libtorch.so)
Nov 19 00:33:35 frame #4: <unknown function> + 0xea99 (0x7fabdd57ba99 in /lib/x86_64-linux-gnu/libpthread.so.0)
Nov 19 00:33:35 frame #5: <unknown function> + 0xb12b363 (0x7fabf034e363 in /var/lib/jenkins/workspace/build/lib/libtorch.so)
Nov 19 00:33:35 frame #6: <unknown function> + 0xea99 (0x7fabdd57ba99 in /lib/x86_64-linux-gnu/libpthread.so.0)
Nov 19 00:33:35 frame #7: <unknown function> + 0xbcffbc (0x56250335cfbc in build/bin/test_api)
Nov 19 00:33:35 frame #8: <unknown function> + 0xbea6f1 (0x5625033776f1 in build/bin/test_api)
Nov 19 00:33:35 frame #9: <unknown function> + 0xbb270e (0x56250333f70e in build/bin/test_api)
Nov 19 00:33:35 frame #10: void test_isfinite<(c10::ScalarType)0, unsigned char>(c10::Device const&) + 0x771 (0x5625037cfe31 in build/bin/test_api)
Nov 19 00:33:35 frame #11: FunctionalTest_isfinite_Test::TestBody() + 0x2f5 (0x5625037b2195 in build/bin/test_api)
Nov 19 00:33:35 frame #12: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x2bb (0x5625040aafdb in build/bin/test_api)
Nov 19 00:33:35 frame #13: testing::Test::Run() + 0x25b (0x562504059acb in build/bin/test_api)
Nov 19 00:33:35 frame #14: testing::TestInfo::Run() + 0x543 (0x56250405c413 in build/bin/test_api)
Nov 19 00:33:35 frame #15: testing::TestCase::Run() + 0x5fb (0x56250405e33b in build/bin/test_api)
Nov 19 00:33:35 frame #16: testing::internal::UnitTestImpl::RunAllTests() + 0x13d5 (0x562504080745 in build/bin/test_api)
Nov 19 00:33:35 frame #17: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x2b2 (0x5625040b0712 in build/bin/test_api)
Nov 19 00:33:35 frame #18: testing::UnitTest::Run() + 0x29a (0x56250407f1ea in build/bin/test_api)
Nov 19 00:33:35 frame #19: main + 0x458 (0x5625032bc328 in build/bin/test_api)
Nov 19 00:33:35 frame #20: __libc_start_main + 0xf0 (0x7fabdc9ea830 in /lib/x86_64-linux-gnu/libc.so.6)
Nov 19 00:33:35 frame #21: _start + 0x29 (0x5625032bbcd9 in build/bin/test_api)
Nov 19 00:33:35 " thrown in the test body.
Nov 19 00:33:35 [  FAILED  ] FunctionalTest.isfinite (12 ms)
```
We might need to split this test into `TEST_F(FunctionalTest, isfinite)` and `TEST_F(FunctionalTest, isfinite_CUDA)`, and the latter would contain only the CUDA tests.